ReplicatedMapTimeToLiveTest AssertionError, Replicated Map should be empty

In current Master 

From  http://54.87.52.100/~jenkins/workspace/Hazelcast-Simulator-nightly/3.6-RC2-SNAPSHOT/2015_11_19-18_55_50/

the test
https://github.com/hazelcast/hazelcast-simulator/blob/master/tests/src/main/java/com/hazelcast/simulator/tests/replicatedmap/ReplicatedMapTimeToLiveTest.java#L85

```
Failure[
  message='Worked ran into an unhandled exception'
  type=WORKER_EXCEPTION
  timestamp=1447956433548
  workerAddress=C_A2_W1
  agentAddress=52.91.158.193
  hzAddress=10.0.0.145:5701
  workerId=worker-52.91.158.193-1-member
  test=TestCase{
       id=ReplicatedTTL,
       class=com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest,
       getProb=0.3,
       putTTLProb=0.7
   }
  cause=java.lang.AssertionError: ReplicatedMapTimeToLiveTest: Replicated Map should be empty, some TTL events are not processed expected:<0> but was:<289>
```

```
./output/2015-11-19__18_58_58/worker-54.164.134.224-10-client/worker.log
[jenkins@ip-10-72-134-107 2015_11_19-18_55_50]$ find . -name worker.log | xargs grep 'assert'
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:51,569 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:51,673 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:51,827 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:52,055 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:52,408 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:52,916 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:53,676 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:54,814 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:56,523 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 17:59:59,078 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:00:02,910 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:00:08,656 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:00:17,273 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:00:30,203 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:00:49,587 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:01:18,666 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:02:02,275 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:03:07,688 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:04:45,804 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
./output/2015-11-19__18_58_58/worker-52.91.158.193-1-member/worker.log:INFO  2015-11-19 18:07:12,980 [Thread-83] com.hazelcast.simulator.tests.replicatedmap.ReplicatedMapTimeToLiveTest: ReplicatedMapTimeToLiveTest: assert map Size = 289
```


unfortunately the `assert map Size = 289` is not decreasing
and the same with the EE run

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapEvictionProcessor.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapEvictionProcessor.java
@@ -24,6 +24,8 @@ import com.hazelcast.util.scheduler.ScheduledEntry;
 import com.hazelcast.util.scheduler.ScheduledEntryProcessor;
 import java.util.Collection;
 
+import static com.hazelcast.replicatedmap.impl.ReplicatedMapService.SERVICE_NAME;
+
 /**
  * Actual eviction processor implementation to remove values to evict values from the replicated map
  */
@@ -42,6 +44,6 @@ public class ReplicatedMapEvictionProcessor implements ScheduledEntryProcessor<O
     public void process(EntryTaskScheduler<Object, Object> scheduler, Collection<ScheduledEntry<Object, Object>> entries) {
         EvictionOperation evictionOperation = new EvictionOperation(store, entries);
         evictionOperation.setPartitionId(partitionId);
-        nodeEngine.getOperationService().executeOperation(evictionOperation);
+        nodeEngine.getOperationService().invokeOnTarget(SERVICE_NAME, evictionOperation, nodeEngine.getThisAddress());
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapService.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapService.java
@@ -80,7 +80,7 @@ public class ReplicatedMapService implements ManagedService, RemoteService, Even
     public static final String SERVICE_NAME = "hz:impl:replicatedMapService";
     public static final int INVOCATION_TRY_COUNT = 3;
 
-    private static final int SYNC_INTERVAL_SECONDS = 10;
+    private static final int SYNC_INTERVAL_SECONDS = 30;
     private static final int MAX_CLEAR_EXECUTION_RETRY = 5;
 
     private final Config config;

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/EvictionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/EvictionOperation.java
@@ -43,11 +43,6 @@ public class EvictionOperation extends AbstractOperation {
     }
 
     @Override
-    public boolean returnsResponse() {
-        return false;
-    }
-
-    @Override
     public boolean validatesTarget() {
         return false;
     }

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/SyncReplicatedMapDataOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/SyncReplicatedMapDataOperation.java
@@ -67,10 +67,10 @@ public class SyncReplicatedMapDataOperation<K, V> extends AbstractOperation {
             if (oldRecord != null) {
                 replicatedRecord.setHits(oldRecord.getHits());
             }
+            newStorage.putInternal(key, replicatedRecord);
             if (record.getTtl() > 0) {
                 store.scheduleTtlEntry(record.getTtl(), key, value);
             }
-            newStorage.putInternal(key, replicatedRecord);
         }
         newStorage.setVersion(version);
         AtomicReference<InternalReplicatedMapStorage<K, V>> storageRef = store.getStorageRef();

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java
@@ -274,6 +274,9 @@ public abstract class AbstractReplicatedRecordStore<K, V> extends AbstractBaseRe
         newRecord.setCreationTime(record.getCreationTime());
         newRecord.setLastAccessTime(record.getLastAccessTime());
         newRecord.setUpdateTime(record.getLastUpdateTime());
+        if (record.getTtl() > 0) {
+            scheduleTtlEntry(record.getTtl(), key, value);
+        }
         getStorage().putInternal(key, newRecord);
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapTtlTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapTtlTest.java
@@ -19,12 +19,28 @@ import org.junit.runner.RunWith;
 public class ReplicatedMapTtlTest extends ReplicatedMapBaseTest {
 
     @Test
-    public void testPutWithTTL() throws Exception {
+    public void testPutWithTTL_withMigration() throws Exception {
+        int nodeCount = 1;
+        int keyCount = 40000;
+        int operationCount = 40000;
+        int threadCount = 15;
+        int ttl = 500;
+        testPutWithTTL(nodeCount, keyCount, operationCount, threadCount, ttl, true);
+    }
+
+    @Test
+    public void testPutWithTTL_withoutMigration() throws Exception {
         int nodeCount = 5;
         int keyCount = 20000;
         int operationCount = 20000;
         int threadCount = 15;
         int ttl = 500;
+        testPutWithTTL(nodeCount, keyCount, operationCount, threadCount, ttl, false);
+
+    }
+
+    private void testPutWithTTL(int nodeCount, int keyCount, int operationCount, int threadCount, int ttl,
+                                boolean causeMigration) throws InterruptedException {
         TimeUnit timeUnit = TimeUnit.MILLISECONDS;
         TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();
         HazelcastInstance[] instances = factory.newInstances(null, nodeCount);
@@ -35,13 +51,20 @@ public class ReplicatedMapTtlTest extends ReplicatedMapBaseTest {
         for (Thread thread : threads) {
             thread.start();
         }
+        HazelcastInstance instance = null;
+        if (causeMigration) {
+            instance = factory.newHazelcastInstance();
+        }
         for (Thread thread : threads) {
             thread.join();
+        }
+        if (causeMigration) {
+            ReplicatedMap<Object, Object> map = instance.getReplicatedMap(mapName);
+            maps.add(map);
         }
         for (ReplicatedMap map : maps) {
             assertSizeEventually(0, map, 60);
         }
-
     }
 
     private Thread[] createThreads(int count, List<ReplicatedMap> maps, ArrayList<Integer> keys,

